src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs

```diff
@@ -673,6 +673,8 @@ private void EmitBlock(BoundBlock block)
             {
                 EmitUninstrumentedBlock(block);
             }
+
+            ReleaseBlockTemps();
```

Comment:
>ReleaseBlockTemps(); [](http://example.com/codeflow?start=12&length=20)

This feels fragile because we cannot assume that this block necessary maps to one syntactically present in source, which was used for "ref safety analysis already checks refs to rvalues cannot escape blocks". We still might not be done with that block.

Reply:
> we cannot assume that this block necessary maps to one syntactically present in source

Why cannot we assume that? To me it seems we don't synthesize BoundBlocks.

Reply:
> Why cannot we assume that? To me it seems we don't synthesize BoundBlocks.

Because we shouldn't be assuming that. There is nothing wrong in introducing a bound block.

Reply:
> There is nothing wrong in introducing a bound block.

This comment suggests otherwise:

https://github.com/dotnet/roslyn/blob/14244f6fcdf77663d7e866ebcdf45e6881a51195/src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml#L1030

Reply:
> This comment suggests otherwise: ...

I am failing to guess what part of the comment you find relevant. Could you quote that part, etc.? 

For example, ```SyntheticBoundNodeFactory``` has a bunch of helpers to synthesize blocks and they are used for one reason or the other. There is also a way to synthesize a block manually. So, even if you haven't found an example, it doesn't mean there isn't one already, or that one won't be introduced in the future.

Reply:
Sorry, I was referring to this line: "BoundBlock specify SCOPE (visibility) of a variable" but that doesn't imply it's wrong to synthesize blocks, you're right.

Do you have suggestions how to best solve this?

I'm thinking if we can assume blocks are never lost during lowering, we could perhaps make sure synthesized blocks are marked as such (e.g., via WasCompilerGenerated) and then we could call ReleaseBlockTemps only for non-compiler-generated blocks.